# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/wr/webrender/src/compositor/sw_compositor.rs
# Commit: 6d98d185a076
# Full Hash: 6d98d185a07684b3538dd0052cb14b67cba38568
# Author: Glenn Watson <git@intuitionlibrary.com>
# Date: 2021-11-05 09:34:21
# Regressor Bug: 1730695
# File Overlap Count: 1
# Description:
#   Bug 1730695 - Fix panic when casting large value to i32 rect r=gfx-reviewers,bradwerth,nical,lsalzman
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D128122
# ==============================================================================

diff -r bba94c79f3e1 -r 6d98d185a076 gfx/wr/webrender/src/compositor/sw_compositor.rs
--- a/gfx/wr/webrender/src/compositor/sw_compositor.rs	Thu Nov 04 19:20:21 2021 +0000
+++ b/gfx/wr/webrender/src/compositor/sw_compositor.rs	Thu Nov 04 19:44:32 2021 +0000
@@ -72,8 +72,8 @@
         clip_rect: &DeviceIntRect,
     ) -> Option<DeviceIntRect> {
         let bounds = self.local_bounds(surface);
-        let device_rect = transform.outer_transformed_box2d(&bounds.to_f32())?.round_out().to_i32();
-        device_rect.intersection(clip_rect)
+        let device_rect = transform.outer_transformed_box2d(&bounds.to_f32())?.round_out();
+        Some(device_rect.intersection(&clip_rect.to_f32())?.to_i32())
     }
 
     /// Determine if the tile's bounds may overlap the dependency rect if it were
@@ -101,7 +101,7 @@
         // Offset the valid rect to the appropriate surface origin.
         let valid = self.local_bounds(surface);
         // The destination rect is the valid rect transformed and then clipped.
-        let dest_rect = transform.outer_transformed_box2d(&valid.to_f32())?.round_out().to_i32();
+        let dest_rect = transform.outer_transformed_box2d(&valid.to_f32())?.round_out().try_cast()?;
         if !dest_rect.intersects(clip_rect) {
             return None;
         }
@@ -157,8 +157,8 @@
         clip_rect: &DeviceIntRect,
     ) -> Option<DeviceIntRect> {
         let bounds = self.local_bounds();
-        let device_rect = transform.outer_transformed_box2d(&bounds.to_f32())?.round_out().to_i32();
-        device_rect.intersection(clip_rect)
+        let device_rect = transform.outer_transformed_box2d(&bounds.to_f32())?.round_out();
+        Some(device_rect.intersection(&clip_rect.to_f32())?.to_i32())
     }
 }
 